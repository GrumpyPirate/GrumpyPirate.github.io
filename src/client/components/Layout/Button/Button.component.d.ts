import { ChildrenProps } from 'types';

interface ButtonProps extends ChildrenProps {
  to?: string;
  external?: boolean;
  primary?: boolean;
  secondary?: boolean;
}

export { ButtonProps };
