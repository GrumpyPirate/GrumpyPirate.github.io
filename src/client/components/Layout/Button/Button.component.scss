@import 'core';

@mixin button-variant($color, $background-color) {
  background-color: $background-color;
  border-color: $background-color;

  &,
  &:hover,
  &:focus,
  &:active {
    color: $color;
  }

  &:hover,
  &:focus {
    background-color: darken($background-color, 5%);
    border-color: darken($background-color, 5%);
  }

  &:active {
    background-color: $background-color;
    border-color: $background-color;
    transform: translateY(1px);
    transition-duration: 0s, $anim-dur, $anim-dur;
  }
}

.button {
  @include button-reset;
  @include no-underline;

  background-color: $btn-primary-bg;
  border: solid 1px $btn-primary-bg;
  border-radius: $btn-border-radius;
  box-shadow: none;
  color: $btn-primary-color;
  display: inline-block;
  font-family: $body-font-family;
  font-size: $btn-font-size;
  font-weight: $btn-font-weight;
  line-height: (rem(22px) / $btn-font-size);
  padding: $btn-padding-y $btn-padding-x;
  text-transform: uppercase;
  transition-duration: $anim-dur;
  transition-property: transform, color, background-color;
  transition-timing-function: $anim-ease;

  &--primary {
    @include button-variant($btn-primary-color, $btn-primary-bg);
  }

  &--secondary {
    @include button-variant($btn-secondary-color, $btn-secondary-bg);
  }

  &:disabled,
  &[disabled] {
    &,
    &:hover,
    &:focus,
    &:active {
      background-color: palette('grey-light');
      border-color: palette('grey-light');
      color: palette('white');
      cursor: not-allowed;
      pointer-events: none;
    }
  }
}
