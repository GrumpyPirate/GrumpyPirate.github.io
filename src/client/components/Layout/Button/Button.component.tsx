import React, { FunctionComponent } from 'react';
import { Link } from 'react-router-dom';
import classnames from 'classnames';

import { ButtonProps } from './Button.component.d';
import classes from './Button.component.scss';

const Button: FunctionComponent<ButtonProps> = ({
  children,
  external,
  primary,
  secondary,
  to = '',
}: ButtonProps) => {
  const isInternalLink = Boolean(to && !external);
  const isExternalLink = Boolean(to && external);
  const classNames = classnames(classes['button'], {
    [classes['button--primary']]: primary,
    [classes['button--secondary']]: secondary,
  });

  if (isInternalLink) {
    return (
      <Link to={to} className={classNames}>
        {children}
      </Link>
    );
  }

  if (isExternalLink) {
    return (
      <a href={to} className={classNames} target="_blank" rel="noreferrer noopener">
        {children}
      </a>
    );
  }

  return (
    <button type="button" className={classNames}>
      {children}
    </button>
  );
};

export default Button;
