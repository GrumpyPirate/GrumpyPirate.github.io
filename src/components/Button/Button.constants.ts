import Color from 'color';
import { css, SimpleInterpolation } from 'styled-components';

import { animDur, palette } from 'styles';

export const btnDisabledStyles = css`
  &:disabled,
  &[disabled] {
    &,
    &:hover,
    &:focus,
    &:active {
      background-color: ${palette.greyLight};
      border-color: ${palette.greyLight};
      color: ${palette.white};
      cursor: not-allowed;
      pointer-events: none;
    }
  }
`;
export const createButtonVariant = (fgColor: string, bgColor: string): SimpleInterpolation => css`
  background-color: ${bgColor};
  border-color: ${bgColor};

  &,
  &:hover,
  &:focus,
  &:active {
    color: ${fgColor};
  }

  &:hover,
  &:focus {
    background-color: ${Color(bgColor).darken(0.05).string()};
    border-color: ${Color(bgColor).darken(0.05).string()};
  }

  &:active {
    background-color: ${bgColor};
    border-color: ${bgColor};
    transform: translateY(1px);
    transition-duration: 0s, ${animDur}s, ${animDur}s;
  }

  ${btnDisabledStyles}
`;
