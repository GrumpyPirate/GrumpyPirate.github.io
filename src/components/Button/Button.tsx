import React, { FunctionComponent } from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';

import {
  animDur,
  animEase,
  bodyFontFamily,
  btnFontSize,
  buttonReset,
  linkFontWeight,
  noUnderline,
  palette,
  rem,
} from 'styles';

import { btnDisabledStyles, createButtonVariant } from './Button.constants';
import { ButtonProps } from './Button.types';

const Button: FunctionComponent<ButtonProps> = function ({
  children,
  className,
  external,
  to = '',
}) {
  const isInternalLink = Boolean(to && !external);
  const isExternalLink = Boolean(to && external);

  if (isInternalLink) {
    return (
      <Link to={to} className={className}>
        {children}
      </Link>
    );
  }

  if (isExternalLink) {
    return (
      <a href={to} className={className} target="_blank" rel="noreferrer noopener">
        {children}
      </a>
    );
  }

  return (
    <button type="button" className={className}>
      {children}
    </button>
  );
};

export default styled(Button)`
  ${buttonReset}
  ${noUnderline}

  background-color: ${palette.themeAccentLight};
  border: solid 1px ${palette.themeAccentLight};
  border-radius: ${rem(btnFontSize * 2)};
  box-shadow: none;
  color: ${palette.white};
  display: inline-block;
  font-family: ${bodyFontFamily};
  font-size: ${rem(btnFontSize)};
  font-weight: ${linkFontWeight};
  line-height: ${22 / btnFontSize};
  padding: ${rem(8)} ${rem(24)};
  text-transform: uppercase;
  transition-duration: ${animDur}s;
  transition-property: transform, color, background-color;
  transition-timing-function: ${animEase};

  ${btnDisabledStyles}

  ${({ primary }) => !!primary && createButtonVariant(palette.white, palette.themeAccentLight)}
  ${({ secondary }) => !!secondary && createButtonVariant(palette.white, palette.themeAccentDark)}
`;
