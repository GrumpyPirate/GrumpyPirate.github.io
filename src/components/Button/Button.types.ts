import { PropsWithChildren } from 'react';

import { ClassNameProps } from 'types/common';

export type ButtonProps = ClassNameProps &
  PropsWithChildren<{
    to?: string;
    external?: boolean;
    primary?: boolean;
    secondary?: boolean;
  }>;
