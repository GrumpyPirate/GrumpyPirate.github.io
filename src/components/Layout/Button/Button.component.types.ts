import { ChildrenProps } from 'types/common';

export interface ButtonProps extends ChildrenProps {
  to?: string;
  external?: boolean;
  primary?: boolean;
  secondary?: boolean;
}
