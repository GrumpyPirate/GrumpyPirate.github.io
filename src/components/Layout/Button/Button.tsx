import classnames from 'classnames';
import React, { FunctionComponent } from 'react';
import { Link } from 'react-router-dom';

import classes from './Button.scss';
import { ButtonProps } from './Button.types';

const Button: FunctionComponent<ButtonProps> = ({
  children,
  external,
  primary,
  secondary,
  to = '',
}) => {
  const isInternalLink = Boolean(to && !external);
  const isExternalLink = Boolean(to && external);
  const classNames = classnames(classes['button'], {
    [classes['button--primary']]: primary,
    [classes['button--secondary']]: secondary,
  });

  if (isInternalLink) {
    return (
      <Link to={to} className={classNames}>
        {children}
      </Link>
    );
  }

  if (isExternalLink) {
    return (
      <a href={to} className={classNames} target="_blank" rel="noreferrer noopener">
        {children}
      </a>
    );
  }

  return (
    <button type="button" className={classNames}>
      {children}
    </button>
  );
};

export default Button;
