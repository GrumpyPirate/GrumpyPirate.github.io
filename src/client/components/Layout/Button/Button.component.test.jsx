import React from 'react';
import { shallow } from 'enzyme';

import Button from './Button.component';

import fixture from './Button.component.fixture';

describe('Components', () => {
  describe('Button', () => {
    const props = {
      ...fixture,
    };

    it('should be defined', () => {
      expect(Button).toBeDefined();
    });

    it('should be a function', () => {
      expect(typeof Button).toBe('function');
    });

    it('should render as expected, without crashing', () => {
      const wrapper = shallow(<Button {...props} />);

      expect(wrapper).toMatchSnapshot();
    });
  });
});
