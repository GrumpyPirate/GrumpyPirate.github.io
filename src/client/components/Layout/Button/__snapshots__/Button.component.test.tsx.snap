// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components Button should render as expected, without crashing 1`] = `
<button
  className="button"
  type="button"
>
  <span>
    Test button
  </span>
</button>
`;
